src/core/meshcop/joiner.cpp

```diff
@@ -75,11 +76,9 @@ void Joiner::GetJoinerId(Mac::ExtAddress &aJoinerId) const
 
 void Joiner::SetState(otJoinerState aState)
 {
-    VerifyOrExit(aState != mState);
+    SuccessOrExit(Get<Notifier>().Update(mState, aState, OT_CHANGED_JOINER_STATE));
 
     otLogInfoMeshCoP("JoinerState: %s -> %s", JoinerStateToString(mState), JoinerStateToString(aState));
```

Comment:
Need to save the old value for logging to be correct. With changing the `mState` value first through `Update()` call, this log will now show the same value for `mState` and `aState`.